repos/system_upgrade/el7toel8/actors/peseventsscanner/libraries/library.py

```diff
@@ -248,31 +235,62 @@ def process_events(events, installed_pkgs):
-            tasks['to_install'].update(non_installed_out_pkgs)
-            # Add the already installed "out" pkgs to to_keep to make sure the repo they're in on RHEL 8 gets enabled
-            installed_out_pkgs = get_installed_out_pkgs(event.out_pkgs, installed_pkgs)
-            tasks['to_keep'].update(installed_out_pkgs)
-            if event.action in ('Split', 'Merged'):
-                # Uninstall those RHEL 7 pkgs that are no longer on RHEL 8
-                in_pkgs_without_out_pkgs = filter_out_out_pkgs(event.in_pkgs, event.out_pkgs)
-                tasks['to_remove'].update(in_pkgs_without_out_pkgs)
-
-        if event.action in ('Renamed', 'Replaced', 'Removed'):
-            # Uninstall those RHEL 7 pkgs that are no longer on RHEL 8
-            tasks['to_remove'].update(event.in_pkgs)
+        'to_remove': {}
+    }
+    for release in releases:
+        slated = {
+            'to_keep': {},
+            'to_install': {},
+            'to_remove': {}
+        }
+        for event in [e for e in events if e.release == release]:
+            if all(
+                (p in installed_pkgs or p in list(tasks['to_install'].keys()))
+                    and p not in list(tasks['to_remove'].keys())
+                    for p in event.in_pkgs.keys()
+            ):
+                if event.action in ('Deprecated', 'Present'):
+                    # Add these packages to to_keep to make sure the repo they're in on RHEL 8 gets enabled
+                    slated['to_keep'].update(event.in_pkgs)
+
+                if event.action == 'Moved':
+                    # Add these packages to to_keep to make sure the repo they're in on RHEL 8 gets enabled
+                    # We don't care about the "in_pkgs" as it contains always just one pkg - the same as the "out" pkg
+                    slated['to_keep'].update(event.out_pkgs)
+
+                if event.action in ('Split', 'Merged', 'Renamed', 'Replaced'):
+                    non_installed_out_pkgs = filter_out_installed_pkgs(event.out_pkgs, installed_pkgs)
+                    slated['to_install'].update(non_installed_out_pkgs)
+                    # Add already installed "out" pkgs to to_keep to ensure the repo they're in on RHEL 8 gets enabled
+                    installed_out_pkgs = get_installed_out_pkgs(event.out_pkgs, installed_pkgs)
+                    slated['to_keep'].update(installed_out_pkgs)
+                    if event.action in ('Split', 'Merged'):
+                        # Uninstall those RHEL 7 pkgs that are no longer on RHEL 8
+                        in_pkgs_without_out_pkgs = filter_out_out_pkgs(event.in_pkgs, event.out_pkgs)
+                        slated['to_remove'].update(in_pkgs_without_out_pkgs)
+
+                if event.action in ('Renamed', 'Replaced', 'Removed'):
+                    # Uninstall those RHEL 7 pkgs that are no longer on RHEL 8
+                    slated['to_remove'].update(event.in_pkgs)
```

Comment:
```suggestion
                    installed_in_pkgs = get_installed_out_pkgs(event.in_pkgs, installed_pkgs)
                    slated['to_remove'].update(installed_in_pkgs)
```
This is to avoid trying to remove what is not installed and thus to reduce the possibility of conflicts.
And please rename the `get_installed_out_pkgs` since it'll not be just about `out` packages.

Reply:
Done.